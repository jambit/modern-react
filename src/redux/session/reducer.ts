import { createReducer } from 'deox';
import { DEFAULT_TOKEN } from '../../api';

const initialState = {
    token: DEFAULT_TOKEN,
};

export type SessionState = typeof initialState;

export const sessionReducer = createReducer(initialState, (handleAction) => []);
