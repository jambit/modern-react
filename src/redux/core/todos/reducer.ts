import { createReducer } from "deox";
import { addTodoSuccess, setTodoCheckedSuccess, removeTodoSuccess, loadTodos } from "./actions";

export interface TodoEntry {
    id: number;
    label: string;
    checked: boolean;
}

const initialState = {
    list: [] as TodoEntry[],
};

export type TodoState = typeof initialState;

export const todosReducer = createReducer(initialState, (handleAction) => [
    handleAction(loadTodos, (state, action) => ({
        ...state,
        list: action.payload.items
    })),
    handleAction(addTodoSuccess, (state, action) => ({
        ...state,
        list: [...state.list, action.payload.item],
    })),
    handleAction(setTodoCheckedSuccess, (state, action) => ({
        ...state,
        list: state.list.map((entry) => {
            if(entry.id === action.payload.id) {
                return {
                    ...entry,
                    checked: action.payload.checked
                };
            }
            return entry;
        })
    })),
    handleAction(removeTodoSuccess, (state, action) => ({
        ...state,
        list: state.list.filter((entry) => entry.id !== action.payload.id)
    }))
]);