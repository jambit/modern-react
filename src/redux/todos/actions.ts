import { createAction } from "deox";
import { TodoEntry } from "./reducer";

export const loadTodos = createAction('TODO/LOAD', (resolve) => (items: TodoEntry[]) => resolve({ items }));
export const addTodoStart = createAction('TODO/ADD/START', (resolve) => (label: string) => resolve({ label }));
export const addTodoSuccess = createAction('TODO/ADD/SUCCESS', (resolve) => (item: TodoEntry) => resolve({ item }));
export const addTodoFailure = createAction('TODO/ADD/FAILURE', (resolve) => (label: string) => resolve({ label }));
export const setTodoCheckedStart = createAction('TODO/SET_CHECKED/START', (resolve) => (id: number, checked: boolean) => resolve({ id, checked }));
export const setTodoCheckedSuccess = createAction('TODO/SET_CHECKED/SUCCESS', (resolve) => (id: number, checked: boolean) => resolve({ id, checked }));
export const setTodoCheckedFailure = createAction('TODO/SET_CHECKED/FAILURE', (resolve) => (id: number, checked: boolean) => resolve({ id, checked }));
export const removeTodoStart = createAction('TODO/REMOVE/START', (resolve) => (id: number) => resolve({ id }));
export const removeTodoSuccess = createAction('TODO/REMOVE/SUCCESS', (resolve) => (id: number) => resolve({ id }));
export const removeTodoFailure = createAction('TODO/REMOVE/FAILURE', (resolve) => (id: number) => resolve({ id }));
