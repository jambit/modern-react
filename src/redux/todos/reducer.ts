import { createReducer } from 'deox';
import { addTodo, setTodoChecked } from './actions';


interface TodoEntry {
    id: number;
    label: string;
    checked: boolean;
}

const initialState = {
    list: [] as TodoEntry[],
    nextId: 0,
};

export type TodoState = typeof initialState;

export const todosReducer = createReducer(initialState, (handleAction) => [
    handleAction(addTodo, (state, action) => ({
        ...state,
        list: [...state.list, { id: state.nextId, label: action.payload.label, checked: false }],
        nextId: state.nextId + 1,
    })),
    handleAction(setTodoChecked, (state, action) => ({
        ...state,
        list: state.list.map((todo) => {
            if (todo.id === action.payload.id) {
                return { ...todo, checked: action.payload.checked };
            }
            return todo;
        })
    })),
]);
