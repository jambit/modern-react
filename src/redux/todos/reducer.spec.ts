import { todosReducer, TodoState, TodoEntry } from './reducer';
import { loadTodos, addTodoSuccess, setTodoCheckedSuccess, removeTodoSuccess } from './actions';

describe('todos reducer', () => {
    const stateWithTwoEntries: TodoState = {
        list: [
            { id: 10, checked: true, label: 'Run errands' },
            { id: 12, checked: false, label: 'Solve Puzzle' },
        ],
    };

    it('returns the initial state when passing in undefined as previous state', () => {
        const newState = todosReducer(undefined, {} as any);

        expect(newState).toEqual({
            list: [],
        });
    });

    describe('loadTodos', () => {
        it('replaces the existing list', () => {
            const newList: TodoEntry[] = [{ id: 5, checked: false, label: 'Buy chocolate' }];
            const newState = todosReducer(stateWithTwoEntries, loadTodos(newList));

            expect(newState).toEqual({
                list: newList,
            });
        });
    });

    describe('addTodoSuccess', () => {
        it('appends an item to the list', () => {
            const entry: TodoEntry = { id: 5, checked: false, label: 'Buy chocolate' };
            const newState = todosReducer(stateWithTwoEntries, addTodoSuccess(entry));

            expect(newState).toEqual({
                list: [...stateWithTwoEntries.list, entry],
            });
        });
    });

    describe('setTodoCheckedSuccess', () => {
        it('appends an item to the list', () => {
            // 1. show snapshot
            // 2. show what happens if changes are done
            // 3. show what happens if changes are done and message changes as well
            const newState = todosReducer(stateWithTwoEntries, setTodoCheckedSuccess(10, false));

            expect(newState).toMatchSnapshot();

            // remind of difference between toBe and toEqual
            expect(newState).not.toBe(stateWithTwoEntries);
            expect(newState.list[0]).not.toBe(stateWithTwoEntries.list[0]);
            expect(newState.list[1]).toBe(stateWithTwoEntries.list[1]);
        });
    });

    describe('removeTodoSuccess', () => {
        it('appends an item to the list', () => {
            const newState = todosReducer(stateWithTwoEntries, removeTodoSuccess(10));

            expect(newState).toEqual({
                list: [stateWithTwoEntries.list[1]],
            });
            expect(newState.list[0]).toBe(stateWithTwoEntries.list[1]);
        });
    });

    // show todos saga next
});
