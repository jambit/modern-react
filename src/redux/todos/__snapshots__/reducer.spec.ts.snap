// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducer setTodoCheckedSuccess appends an item to the list 1`] = `
Object {
  "list": Array [
    Object {
      "checked": false,
      "id": 10,
      "label": "Run errands",
    },
    Object {
      "checked": false,
      "id": 12,
      "label": "Solve Puzzle",
    },
  ],
}
`;
